src/Http/WebUtilities/src/FormPipeReader.cs

```diff
@@ -217,53 +222,68 @@ public FormPipeReader(PipeReader pipeReader, Encoding encoding)
-                    if ((sequenceReader.Consumed - consumedBytes) > KeyLengthLimit)
+                    if (!isFinalBlock)
                     {
-                        ThrowKeyTooLargeException();
+                        // Don't buffer indefinately
+                        if ((sequenceReader.Consumed - consumedBytes) > KeyLengthLimit + ValueLengthLimit)
+                        {
+                            ThrowKeyOrValueTooLargeException();
+                        }
+                        break;
                     }
 
-                    break;
+                    // This must be the final key=value pair
+                    segment = buffer.Slice(sequenceReader.Position);
+                    sequenceReader.Advance(segment.Length);
                 }
 
-                if (key.Length > KeyLengthLimit)
+                if (segment.IsSingleSegment)
                 {
-                    ThrowKeyTooLargeException();
+                    ParseFormValuesFast(segment.First.Span, ref accumulator, isFinalBlock: true, out var segmentConsumed);
+                    Debug.Assert(segmentConsumed == segment.First.Span.Length);
+                    continue;
                 }
 
-                if (!sequenceReader.TryReadTo(out ReadOnlySequence<byte> value, andDelimiter, false) ||
-                    !sequenceReader.IsNext(andDelimiter, true))
+                var segmentReader = new SequenceReader<byte>(segment);
+                ReadOnlySequence<byte> value;
+
+                if (segmentReader.TryReadTo(out var key, equalsDelimiter))
                 {
-                    if (!isFinalBlock)
+                    if (key.Length > KeyLengthLimit)
                     {
-                        if ((sequenceReader.Consumed - consumedBytes - key.Length) > ValueLengthLimit)
-                        {
-                            ThrowValueTooLargeException();
-                        }
-                        break;
+                        ThrowKeyTooLargeException();
                     }
 
-                    value = buffer.Slice(sequenceReader.Position);
-
-                    sequenceReader.Advance(value.Length);
+                    value = segment.Slice(segmentReader.Position);
+                    if (value.Length > ValueLengthLimit)
+                    {
+                        ThrowValueTooLargeException();
+                    }
                 }
-
-                if (value.Length > ValueLengthLimit)
+                else
                 {
-                    ThrowValueTooLargeException();
+                    // Too long for the whole segment to be a key.
+                    if (segment.Length > KeyLengthLimit)
+                    {
+                        ThrowKeyTooLargeException();
+                    }
+
+                    // There is no more data, this segment must be "key" with no equals or value.
+                    key = segment;
+                    value = ReadOnlySequence<byte>.Empty;
```

Comment:
Do you need to assign this?

Reply:
No. It's here for clarity and consistency with the fast path.

Reply:
Actually I do need to set it or it complains about using an uninitialized field later.